### ISSUE TITLE
slow vdisk rollback 

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-Disk#470]
As reported by QA, rollback is super slow on packet.net.

I have profiled and try to fix the issue using 3 VMs with following setup:
- 1st VM hosts ardb servers for nbd server
- 2nd VM hosts
      - 1 etcd server
      -  3 0-stor for tlog server : 2 data, 1 parity
- 3rd  hosts
     - tlog and nbdserver, needed by `zeroctl copy` 
     - execute `zeroctl restore`

I have tested `zeroctl copy` but not check it further because:
- it was succeed in QA test
- current zeroctl copy only considered temporary fix

1st & 2nd VM are on digital ocean singapore
I use two kind of 3rd VM:
   - in digital ocean singapore : ping time to 1st & 2nd vm is under < 1ms
   - in vultr  singapore  : ping time to 1st & 2nd vm ~1.3ms

I use xenial-server-cloudimg-amd64-disk1.img as image, it resulted in 232345 tlog transactions, with 4k block size, it is around 800mb

# note about my setup

- all digital ocean VMs have 2 CPU and 2GB of RAM
- vultr VM has 2 CPU and 4GB of RAM
- hosts all ardb in single VM will make it slower, but it is good for this test.
- hosts etcd and 0-stor in single server will make it slower, but it is good for this test

# Test result

## summary of result
- 232345 tlog transactions, with 4k block size, it is around 800mb
- restore code has been optimized, almost 3x faster . From 11 mins to 4 mins 9 sec
- zerotier make the process ~1.5x slower
### copy
```
root@nbtl:~/pathgo/src/github.com/zero-os/0-Disk/bin# time ./zeroctl copy vdisk  myimg copiedimg  --config ~/konfig.yaml --data-shards 2 --parity-shards 1 --priv-key 12345678901234567890123456789012
INFO[09-12|21:42:18] parsing positional arguments...          module=global caller=vdisk.go:53
INFO[09-12|21:42:18] generating tlog data for target vdisk `copiedimg` module=global caller=vdisk.go:113
INFO[09-12|21:42:20] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-12|21:48:10] GenerateFromStorage generates `232345` tlog data with err = <nil> module=global caller=generator.go:177
INFO[09-12|21:48:10] dumping vdisk "myimg" and restoring it as vdisk "copiedimg" module=global caller=deduped.go:369
INFO[09-12|21:48:11] copied 1836 meta indices to vdisk "copiedimg" module=global caller=deduped.go:374

real    5m53.140s
user    3m4.748s
sys     1m10.240s
```

### restore 
```
root@nbtl:~/pathgo/src/github.com/zero-os/0-Disk/bin# ./zeroctl restore vdisk copiedimg --config ~/konfig.yaml --data-shards=2 --parity-shards=1
INFO[09-12|21:59:09] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-12|21:59:10] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-12|22:10:10] restore finished with last sequence = 232344 module=global caller=vdisk.go:84
```

The restore took 11 mins

profile : 
[restore-cpu-prof.pdf](https://github.com/zero-os/0-Disk/files/1298239/restore-cpu-prof.pdf)

# optimization

I found that tlog player could be optimized, by reducing number of block storage flush, resulted in 3d88a87a697072273461f2716988dc2cdfac6188

### restore  with optimization

from DO using public IP
```
root@nbtl:~/pathgo/src/github.com/zero-os/0-Disk/bin# time ./zeroctl restore vdisk copiedimg --config ~/konfig.yaml --data-shards=2 --parity-shards=1
INFO[09-13|03:08:19] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|03:08:20] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|03:12:28] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    4m9.113s
user    1m36.816s
sys     1m10.392s

```
profile: 
[late-flush-restore-cpu-prof.pdf](https://github.com/zero-os/0-Disk/files/1298241/late-flush-restore-cpu-prof.pdf)

from VULTR using public IP
```
root@nbdtlog2:~# time ./zeroctl restore vdisk copiedimg --config ~/konfig.yaml --data-shards=2 --parity-shards=1
INFO[09-13|03:34:14] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|03:34:16] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|03:40:49] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    6m35.625s
user    0m51.264s
sys     0m42.932s
```

[vultr-late-flush-restore-cpu-prof.pdf](https://github.com/zero-os/0-Disk/files/1298242/vultr-late-flush-restore-cpu-prof.pdf)

From DO using zerotier IP
```
root@nbtl:~/pathgo/src/github.com/zero-os/0-Disk/bin# time ./zeroctl restore vdisk copiedimg --config ~/zt_config.yaml --data-shards=2 --parity-shards=1
INFO[09-13|03:44:31] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|03:44:32] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|03:50:39] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    6m8.188s
user    1m32.192s
sys     1m12.132s

root@nbtl:~/pathgo/src/github.com/zero-os/0-Disk/bin# time ./zeroctl restore vdisk copiedimg --config ~/zt_config.yaml --data-shards=2 --parity-shards=1
INFO[09-13|04:17:17] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|04:17:18] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|04:23:15] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    5m58.153s
user    1m28.884s
sys     1m9.172s
```

[zerotier-late-flush-restore-cpu-prof.pdf](https://github.com/zero-os/0-Disk/files/1298243/zerotier-late-flush-restore-cpu-prof.pdf)

From vultr with zerotier IP
```
root@nbdtlog2:~# time ./zeroctl restore vdisk copiedimg --config ~/zt_config.yaml --data-shards=2 --parity-shards=1
INFO[09-13|03:52:28] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|03:52:29] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|04:00:32] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    8m4.846s
user    0m52.440s
sys     0m40.888s

root@nbdtlog2:~# time ./zeroctl restore vdisk copiedimg --config ~/zt_config.yaml --data-shards=2 --parity-shards=1
INFO[09-13|04:27:08] LBACacheLimit (0) will be defaulted to 4096 (min-capped) module=global caller=deduped.go:24
INFO[09-13|04:27:08] restoring vdisk with start timestamp=0 end timestamp=0 module=global caller=vdisk.go:82
INFO[09-13|04:35:20] restore finished with last sequence = 232344,err = <nil> module=global caller=vdisk.go:84

real    8m12.004s
user    0m54.252s
sys     0m43.588s

```
[zerotier-vultr-late-flush-restore-cpu-prof.pdf](https://github.com/zero-os/0-Disk/files/1298245/zerotier-vultr-late-flush-restore-cpu-prof.pdf)

**Comment by islamTaha12** [comment]
I think this issue is duplicate to 
https://github.com/zero-os/0-orchestrator/issues/808

**Comment by iwanbk** [comment]
> I think this issue is duplicate to
> zero-os/0-orchestrator#808

I think the duplicate one is #425, but it is OK to be duplicated in separate repository. Details here are from 0-Disk point of view.,

**Comment by GlenDC** [comment]
@iwanbk what do you want to do with this issue? Are you happy with your fix for now and close it with commit 3d88a87, or do you want to move it to beta-1 and continue working on it?

**Comment by iwanbk** [comment]
closing it
